### PR TITLE
feat: All outgoing requests should go through the configured proxy

### LINKED ISSUE #11749: [Feature]: Add support for Appsmith to connect through a http or https proxy server
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

When the user self hosts Appsmith in their network, there is a very high chance that all network calls to the outside world will be blocked. Usually the outside/internet calls are done via proxy server which is used only for the purpose of interacting with the outside world. 

This feature lets the users to use features which are dependent on Cloud Service such as OAuth, GitSync, Licencing etc without them having to whitelist ip/domain of Cloud Service. 

### Why should this be worked on?

We use cloud service for a lot of activities now a day. When the user self hosts appsmith in their network, there is a very high chance that all network calls to the outside world will be blocked. Usually the outside/internet calls are done via proxy server which is used only for the purpose of interacting with the outside world. With this support we can support such requirements from users.

### PR DESCRIPTION
**Pull request by sharat87**
Fixes #11749

## To test

1. Install `mitmproxy` with `brew install mitmproxy`.
2. Start the proxy with `mitmweb --listen-port 9020 --web-port 9021`.
4. Start Appmith with:
   ```
   docker run --name ace -p 8001:80 -v ~/appsmith-stacks:/appsmith-stacks -d --pull always -e HTTP_PROXY=http://host.docker.internal:9020 -e HTTPS_PROXY=http://host.docker.internal:9020 appsmith/appsmith-ce:release
   ```
5. Now launch Appsmith at http://localhost:8001, load the templates page, and check that a request to load templates from cloud-services shows up in the proxy's web UI.
6. Run any API request in Appsmith and see that that also appears in the proxy's web UI.
7. Try connect a git repo and check that git related queries are also made via the proxy.

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/EnvManagerCEImpl.java [hunk 1]
```diff
@@ -159,12 +159,54 @@ public List<String> transformEnvContent(String envContent, Map<String, String> c
...
 
         for (final String name : remainingChangedNames) {
-            outLines.add(name + "=" + changes.get(name));
+            outLines.add(name + "='" + changes.get(name).replace("'", "'\\''") + "'");
         }
 
         return outLines;
     }
 
+    /*
```
Comment (trishaanand) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/EnvManagerCEImpl.java:168]: @sharat87 : Left this change by mistake?

Comment (sharat87) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/EnvManagerCEImpl.java:168]: Of course I did. I really need to stop doing multiple things at once. 😖

Comment (sharat87) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/EnvManagerCEImpl.java]: Resolved, thanks for the find.

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/util/WebClientUtils.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+                .baseUrl(baseUrl)
+                .build();
+    }
+
+    private static boolean shouldUseSystemProxy() {
+        return "true".equals(System.getProperty("java.net.useSystemProxies"))
+                && (!System.getProperty("http.proxyHost", "").isEmpty() || !System.getProperty("https.proxyHost", "").isEmpty());
+    }
+
+    public static WebClient.Builder builder() {
```
Comment (nidhi-nair) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/util/WebClientUtils.java:28]: Hey why have we left REST API out of this? @sharat87

Comment (sharat87) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/util/WebClientUtils.java:28]: Rest API plugin also uses the same piece of code, but is duplicated. I couldn't combine the two APIs to work togather, and have put it aside. Let me see if I can do it without making the implementation too messy.

Comment (sharat87) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/util/WebClientUtils.java:28]: Done, thanks. :)

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/util/WebClientUtils.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        HttpClient httpClient = applyProxyIfConfigured(HttpClient.create());
+
+        if (shouldUseSystemProxy()) {
+            httpClient = httpClient.proxyWithSystemProperties();
+        }
+
+        return builder(httpClient);
+    }
+
+    public static WebClient.Builder builder(HttpClient httpClient) {
```
Comment (nidhi-nair) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/util/WebClientUtils.java:38]: I might be missing something here but this code path does not seem to go through the `applyProxyIfConfigured` method right? @sharat87

Comment (sharat87) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/util/WebClientUtils.java:38]: Yep, good catch! Fixed now. :)

**Review by nidhi-nair — APPROVED** [review]
LGTM

**Comment by sharat87** [comment]
@nidhi-nair, one last review/approval please? :D

**Review by nidhi-nair — APPROVED** [review]
Woohoo!! 🎉🎉🎉

**Pull request merged**